### PR TITLE
Moved `get_component(_unchecked_mut)` from `Query` to `QueryState`

### LINKED ISSUE #9683: Add `QueryState::(get_)component(_mut)`
## What problem does this solve or what need does it fill?

`Query` have these methods, but `QueryState` - doesn't.

## What solution would you like?

Add these methods that call corresponding methods from `Query`.

### PR DESCRIPTION
**Pull request by bushrat011899**
# Objective

- Fixes #9683

## Solution

- Moved `get_component` from `Query` to `QueryState`.
- Moved `get_component_unchecked_mut` from `Query` to `QueryState`.
- Moved `QueryComponentError` from `bevy_ecs::system` to `bevy_ecs::query`. Minor Breaking Change.
- Narrowed scope of `unsafe` blocks in `Query` methods.

---

## Migration Guide

- `use bevy_ecs::system::QueryComponentError;` -> `use bevy_ecs::query::QueryComponentError;`

## Notes

I am not very familiar with unsafe Rust nor its use within Bevy, so I may have committed a Rust faux pas during the migration.

### CODE REVIEW

File: crates/bevy_ecs/src/system/query.rs [hunk 1]
```diff
@@ -1401,13 +1343,13 @@ impl<'w, 's, Q: WorldQuery, F: ReadOnlyWorldQuery> Query<'w, 's, Q, F> {
...
-            self.state
-                .as_nop()
-                .get_unchecked_manual(self.world, entity, self.last_run, self.this_run)
-                .is_ok()
-        }
+        let state = self.state.as_nop();
+
+        let result =
+            // SAFETY: NopFetch does not access any members while &self ensures no one has exclusive access
+            unsafe { state.get_unchecked_manual(self.world, entity, self.last_run, self.this_run) };
```
Comment (tguichaoua) [inline crates/bevy_ecs/src/system/query.rs:1350]:
```suggestion
        // SAFETY: NopFetch does not access any members while &self ensures no one has exclusive access
        let result = unsafe { state.get_unchecked_manual(self.world, entity, self.last_run, self.this_run) };
```

It's hard to read when the `SAFETY` comment is in the middle of the statement

File: crates/bevy_ecs/src/system/query.rs [hunk 2]
```diff
@@ -1172,33 +1136,15 @@ impl<'w, 's, Q: WorldQuery, F: ReadOnlyWorldQuery> Query<'w, 's, Q, F> {
...
-            .has_write(archetype_component)
-        {
-            entity_ref
-                .get_mut_using_ticks::<T>(self.last_run, self.this_run)
-                .ok_or(QueryComponentError::MissingComponent)
-        } else {
-            Err(QueryComponentError::MissingWriteAccess)
+
+        // SAFETY: this check is required to ensure soundness in the case of `to_readonly().get_component_mut()`
+        // See the comments on the `force_read_only_component_access` field for more info.
```
Comment (tguichaoua) [inline crates/bevy_ecs/src/system/query.rs:1144]: This comment should be kept above `if self.force_read_only_component_access {` since it describes why the check is needed.

A more suitable comment for this unsafe block could be something like `SAFETY: requirements are uphold by the caller`

**Review by tguichaoua — CHANGES_REQUESTED** [review]

**Comment by Shatur** [comment]
Does it makes sense to move `component` and `component_mut` too?

**Comment by bushrat011899** [comment]
> Does it makes sense to move `component` and `component_mut` too?

It might do, I started with `get_component` and `get_component_unchecked_mut` because they were the last two methods that had sizeable implementations. There are several other functions which may want to be reworked as well:

* `par_iter`: This doesn't require any additional information beyond `get_component_unchecked_mut`.
* `par_iter_mut`: See `par_iter`.
* `many`: Currently calls `Query::get_many`; should instead call `QueryState::get_many_read_only_manual`.
* `many_mut`: See `many`.
* `get_component_mut`: See `many`.
* `single`: See `many`.
* `single_mut`: See `many`.
* `contains`: See `par_iter`.

Additionally, neither `Query` nor `QueryState` have the following methods, but probably should for completeness:

* `component`
* `component_mut`

If we also migrate the above to `QueryState`, then all of `Query`'s methods simply forward to `QueryState`. Is that a desirable state for `Query` and `QueryState` to be in?

**Comment by Shatur** [comment]
> Additionally, neither Query nor QueryState have the following methods, but probably should for completeness:

`Query` have these methods since #9659. This is why I asking.

**Comment by bushrat011899** [comment]
> > Additionally, neither Query nor QueryState have the following methods, but probably should for completeness:
> 
> `Query` have these methods since #9659. This is why I asking.

I missed that merge from my branch, apologies! Below would be the complete list of candidates to migrate over:

* `par_iter`: This doesn't require any additional information beyond `get_component_unchecked_mut`.
* `par_iter_mut`: See `par_iter`.
* `many`: Currently calls `Query::get_many`; should instead call `QueryState::get_many_read_only_manual`.
* `many_mut`: See `many`.
* `get_component_mut`: See `many`.
* `single`: See `many`.
* `single_mut`: See `many`.
* `contains`: See `par_iter`.
* `component`: See `many`.
* `component_mut`: See `many`.

I'm going to update my PR with the requested changes from @tguichaoua, then I'll migrate the above over just to demonstrate what that change would look like.

File: crates/bevy_ecs/src/system/query.rs (hunk 1 above)
Comment (bushrat011899) [reply crates/bevy_ecs/src/system/query.rs:1350]: This isn't valid as-is due to `cargo fmt` rules, but I'll destruct the parameters of this function which should get it short enough to get the desired effect.

**Comment by alice-i-cecile** [comment]
> If we also migrate the above to QueryState, then all of Query's methods simply forward to QueryState. Is that a desirable state for Query and QueryState to be in?

Yes; I much prefer that! Unless there's a very good (and commented) reason not to, these APIs should be mirrors of each other. Due to the fact that every `Query` stores a `QueryState`, method forwarding like this is the best way to achieve that.

File: crates/bevy_ecs/src/query/state.rs [hunk 3]
```diff
@@ -1343,6 +1527,99 @@ impl fmt::Display for QueryEntityError {
     }
 }
 
+/// An error that occurs when retrieving a specific [`Entity`]'s component from a [`Query`](crate::system::Query).
+#[derive(Debug, PartialEq, Eq)]
+pub enum QueryComponentError {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/state.rs:1532]: Should this live in `query/mod.rs` instead? I like making it pub though.

File: crates/bevy_ecs/src/system/query.rs [hunk 4]
```diff
@@ -425,14 +426,12 @@ impl<'w, 's, Q: WorldQuery, F: ReadOnlyWorldQuery> Query<'w, 's, Q, F> {
     /// - [`for_each`](Self::for_each) for the closure based alternative.
     #[inline]
     pub fn iter(&self) -> QueryIter<'_, 's, Q::ReadOnly, F::ReadOnly> {
+        let state = self.state.as_readonly();
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/system/query.rs:429]: I think that this set of changes is a bit of a readability regression, and makes the diff a lot noisier.

File: crates/bevy_ecs/src/query/state.rs (hunk 3 above)
Comment (bushrat011899) [reply crates/bevy_ecs/src/query/state.rs:1532]: I think it makes sense to live wherever `QueryEntityError` lives, which is currently `query/state.rs`. Moving both into `query/mod.rs` (or a separate `query/error.rs`) would be my preference.

File: crates/bevy_ecs/src/system/query.rs (hunk 4 above)
Comment (bushrat011899) [reply crates/bevy_ecs/src/system/query.rs:429]: That is a fair critique. My goal with those changes was to minimise the amount of operations happening within an `unsafe` block. I'll revert those changes because they aren't really relevant to the rest of the PR.

File: crates/bevy_ecs/src/query/state.rs (hunk 3 above)
Comment (alice-i-cecile) [reply crates/bevy_ecs/src/query/state.rs:1532]: Let's move them both then.

**Review by alice-i-cecile — APPROVED** [review]

File: crates/bevy_ecs/src/query/state.rs [hunk 5]
```diff
@@ -538,6 +678,28 @@ impl<Q: WorldQuery, F: ReadOnlyWorldQuery> QueryState<Q, F> {
...
+    /// use `QueryState::validate_world` to verify this.
+    pub(crate) unsafe fn many_read_only_manual<'w, const N: usize>(
+        &self,
+        world: &'w World,
+        entities: [Entity; N],
+        last_run: Tick,
+        this_run: Tick,
+    ) -> [ROQueryItem<'w, Q>; N] {
+        self.get_many_read_only_manual(world, entities, last_run, this_run)
+            .unwrap()
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/query/state.rs:700]: Using `unwrap` won't have a good error message, since the error's `Debug` implementation will be used. It'd be better to use a custom panic message or the `Display` impl.

Comment (bushrat011899) [reply crates/bevy_ecs/src/query/state.rs]: I've updated the PR to use a `match` statement to extract the error, and then `panic!` with a short context message, and then the `Display` of the inner error.

```rust
match self.get_many_unchecked_manual(world, entities, last_run, this_run) {
    Ok(items) => items,
    Err(error) => panic!("Cannot get query result: {error}"),
}
```
I noticed there were multiple places where the above pattern was appropriate, so I've implemented it in all such locations within `state.rs`. I also moved `QuerySingleError` into the new `error.rs` module for consistency.

File: crates/bevy_ecs/src/query/state.rs [hunk 6]
```diff
@@ -506,6 +510,142 @@ impl<Q: WorldQuery, F: ReadOnlyWorldQuery> QueryState<Q, F> {
...
+            .ok_or(QueryComponentError::MissingComponent)?;
+        let archetype_component = entity_ref
+            .archetype()
+            .get_archetype_component_id(component_id)
+            .ok_or(QueryComponentError::MissingComponent)?;
+        if self
+            .archetype_component_access
+            .has_write(archetype_component)
+        {
+            // SEMI-SAFETY: It is the responsibility of the caller to ensure it is sound to get a
```
Comment (tguichaoua) [inline crates/bevy_ecs/src/query/state.rs:607]:
```suggestion
            // SAFETY: It is the responsibility of the caller to ensure it is sound to get a
```

File: crates/bevy_ecs/src/system/query.rs [hunk 7]
```diff
@@ -1222,33 +1198,47 @@ impl<'w, 's, Q: WorldQuery, F: ReadOnlyWorldQuery> Query<'w, 's, Q, F> {
...
+    /// # Safety
+    ///
+    /// This function makes it possible to violate Rust's aliasing guarantees.
+    /// You must make sure this call does not result in multiple mutable references to the same component.
+    ///
+    /// # See also
+    ///
+    /// - [`get_component_mut`](Self::get_component_mut) for the safe version.
+    #[inline]
+    pub unsafe fn component_unchecked_mut<T: Component>(&self, entity: Entity) -> Mut<'_, T> {
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/system/query.rs:1230]: I don't think we should add a convenience function for the `unchecked` variant. This doesn't allow you to do anything that `get_component_unchecked_mut` can't do, and we should try to keep our unsafe APIs as small as possible.

Comment (bushrat011899) [reply crates/bevy_ecs/src/system/query.rs:1230]: That seems reasonable, I'll make that change.

File: crates/bevy_ecs/src/query/state.rs [hunk 8]
```diff
@@ -506,6 +510,142 @@ impl<Q: WorldQuery, F: ReadOnlyWorldQuery> QueryState<Q, F> {
...
+    /// # Panics
+    ///
+    /// Panics in case of a nonexisting entity or mismatched component.
+    ///
+    /// # Safety
+    ///
+    /// This function makes it possible to violate Rust's aliasing guarantees.
+    /// You must make sure this call does not result in multiple mutable references to the same component.
+    #[inline]
+    pub(crate) unsafe fn component_unchecked_mut<'w, 's, 'r, T: Component>(
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/query/state.rs:628]: One last thing -- I don't think this method needs to exist anymore. It's private and only called in one place. Leaving it here is just begging someone to make it `pub` in the future, which I don't think we should do for the same reasons I talked about for the `Query` version of this method.

Otherwise, this PR looks good to me.

Comment (bushrat011899) [reply crates/bevy_ecs/src/query/state.rs:628]: I think I disagree on that reasoning, because `many_read_only_manual`, `single`, `single_mut`, `many_unchecked_manual`, etc. all also fit that description. The downside to that change as well is the error handling in those functions then just gets shuffled back into `Query`, which I think is contrary to the goal of this PR, which is to move implementation details out of `Query` and into `QueryState`.

If you still think it should be removed I'm happy to make that change, but I think it's worth pointing out that possible inconsistency.

Comment (JoJoJet) [reply crates/bevy_ecs/src/query/state.rs:628]: It's fine to add panicking variants for `single{_mut}` and `component{_mut}` since those are safe functions. I'd even argue that they should be made public. Good point about `many_*` though -- I'd also argue that they don't need to exist, since they're private, and each only gets called once. There's not much gain to adding more unsafe functions like this IMO.

Ultimately though it's up to you. I'm just sharing my opinion :)

Comment (bushrat011899) [reply crates/bevy_ecs/src/query/state.rs]: And thanks for sharing! I've removed `component_unchecked_mut`, `many_read_only_manual`, and `many_unchecked_manual` as they were only called by `Query` in a single location, and were all `unsafe`. There are possibly more `unsafe` functions in `QueryState` that could be marked as `safe`, but I'm not familiar enough with Bevy or `unsafe` Rust to make that call.

**Review by tguichaoua — APPROVED** [review]

File: crates/bevy_ecs/src/system/query.rs [hunk 9]
```diff
@@ -1194,7 +1171,10 @@ impl<'w, 's, Q: WorldQuery, F: ReadOnlyWorldQuery> Query<'w, 's, Q, F> {
     #[inline]
     #[track_caller]
     pub fn component_mut<T: Component>(&mut self, entity: Entity) -> Mut<'_, T> {
-        match self.get_component_mut(entity) {
+        // SAFETY: unique access to query (preventing aliased access)
+        let result = unsafe { self.get_component_unchecked_mut(entity) };
+
+        match result {
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/system/query.rs:1177]: Hold up -- why was this changed to use `get_component_unchecked_mut`? Can we go back to using the safe method?

**Review by JoJoJet — CHANGES_REQUESTED** [review]

File: crates/bevy_ecs/src/system/query.rs (hunk 9 above)
Comment (bushrat011899) [reply crates/bevy_ecs/src/system/query.rs]: Yeah that makes more sense. I think at the time I thought it made more sense to directly access the `QueryState` version of that function. Fixed now!

**Review by JoJoJet — APPROVED** [review]

File: crates/bevy_ecs/src/query/state.rs [hunk 10]
```diff
@@ -1196,7 +1304,10 @@ impl<Q: WorldQuery, F: ReadOnlyWorldQuery> QueryState<Q, F> {
     #[track_caller]
     #[inline]
     pub fn single<'w>(&mut self, world: &'w World) -> ROQueryItem<'w, Q> {
-        self.get_single(world).unwrap()
+        match self.get_single(world) {
+            Ok(items) => items,
+            Err(error) => panic!("Cannot get single mutable query result: {error}"),
+        }
```
Comment (mnmaita) [inline crates/bevy_ecs/src/query/state.rs:1310]:
```rs
        self.get_single(world).expect(format!("Cannot get single mutable query result: {error}"))
```

I think you can achieve the same by using the `expect()` method, so you can avoid such verbose match block for simply panicking on error. Same thing for the rest of the files.

I used format macro here but I think you could also use string concatenation or whatever works best to display the full error.

Comment (JoJoJet) [reply crates/bevy_ecs/src/query/state.rs:1310]: That would be less performant, since it would do the string formatting regardless of whether an error occurred.

Comment (bushrat011899) [reply crates/bevy_ecs/src/query/state.rs:1310]: I don't believe that's the case, because I'm trying to extract both the `Err(error)` and `Ok(item)` variants. To my knowledge, `.expect(...)` doesn't use the `Display` implementation for the error message with a custom format. If you check this [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2b34b7f9bd813fb3eeb1a12d9a4306f3), you can see the display implementation isn't used, which was specifically requested behaviour from an earlier review.

Comment (mnmaita) [reply crates/bevy_ecs/src/query/state.rs:1310]: Disregard then! I wasn't considering those very valid points and I probably missed that earlier review too.
